pageserver/src/tenant/mgr.rs

```diff
@@ -575,8 +573,9 @@ where
             Some(tenant) => match tenant.current_state() {
                 TenantState::Attaching
                 | TenantState::Loading
+                | TenantState::Activating
                 | TenantState::Broken { .. }
-                | TenantState::Active => tenant.set_stopping(),
+                | TenantState::Active => tenant.set_stopping().await,
```

Comment:
Should this be pulled out of holding TENANTS lock?

Reply:
It's needed for the `wait_for`.

Reply:
> needed for the `wait_for`

What do you mean? 

`set_stopping(&self)`, so the tenant could be smuggled out of this block, and then the await done outside of the write lock on TENANTS.

Reply:
> What do you mean as in how is it needed? I did read the comment above, but it seems this is at most a proxy for request coalescing. The proxy is just a global RWLock :)
> 
> set_stopping(&self), so the tenant could be smuggled out of this block, and then the await done outside of the write lock on TENANTS, pick the one who manages to make the state transition as the one and lucky?
> 

Hm, I can't really follow, maybe best to leave these two paragraphs alone and focus on the third.

> Well, perhaps this could be in a FIXME just as well. I am just worrying that the wait_for might loop for a long time in some unforseen situation.

Let's look at the remaining callers of `set_stopping` and `set_broken`. 

* `shutdown_all_tenants -> set_stopping` => implicit timeout through the SIGKILL. It might make sense to do FuturesUnordered there? WDYT?
* `remove_tenant_from_memory` => this is detach, which currently has cancel-safety prolems[^1]. I think the `wait_for` doesn't change anything here.
* `remove_tenant_from_memory` => `set_broken` is only used if the cleanup future fails, and, it's the last call on the way out. So, I don't expect breakage.


[^1]: can't find the issue, but, it's the one where if client hangs up during detach, we fail because tenant is already `Stopping`

Reply:
Holding write lock on TENANTS map will prevent new page_service connections from being established (globally, not only for this particular tenant), which I think is undesirable

Reply:
fixed in https://github.com/neondatabase/neon/pull/4321/commits/e7c4ef9f4f8ac3e5e5a383839ba0b365e4e6039b and preceding commit

Reply:
I still find the whole `TENANTS.write()` locking questionable, but its difficult to say if it's a load bearing exclusive lock. Resolving anyways.